contribs/gmf/src/directives/search.js

```diff
@@ -0,0 +1,280 @@
+goog.require('ol.Map');
+goog.require('ol.View');
+goog.require('ol.layer.Tile');
+goog.require('ol.layer.Vector');
+goog.require('ol.proj');
+goog.require('ol.source.OSM');
+goog.require('ol.source.Vector');
+
+
+/**
+ * @return {angular.Directive} The Directive Definition Object.
+ * @ngInject
+ */
+gmf.searchDirective = function() {
+  return {
+    restrict: 'E',
+    scope: {
+      'getMapFn': '&gmfSearchMap',
+      'getDatasourcesFn': '&gmfSearchDatasources'
+    },
+    controller: 'GmfSearchController',
+    controllerAs: 'ctrl',
+    template:
+        '<input type="text" placeholder="search…" ' +
+        'ngeo-search="ctrl.options" ' +
+        'ngeo-search-datasets="ctrl.datasets" ' +
+        'ngeo-search-listeners="ctrl.listeners">',
+    link:
+        /**
+         * @param {angular.Scope} scope Scope.
+         * @param {angular.JQLite} element Element.
+         * @param {angular.Attributes} attrs Atttributes.
+         */
+        function(scope, element, attrs) {
+          // Empty the search field on focus and blur.
+          element.find('input').on('focus blur', function() {
+            $(this).val('');
+          });
+        }
+  };
+};
+
+gmfModule.directive('gmfSearch', gmf.searchDirective);
+
+
+
+/**
+ * @constructor
+ * @param {angular.Scope} $scope The directive's scope.
+ * @param {angular.$compile} $compile Angular compile service.
+ * @param {ngeo.CreateGeoJSONBloodhound} ngeoCreateGeoJSONBloodhound The ngeo
+ * create GeoJSON Bloodhound service.
+ * @ngInject
+ * @export
+ */
+gmf.SearchController = function($scope, $compile,
+    ngeoCreateGeoJSONBloodhound) {
+
+  /**
+   * @type {angular.Scope}
+   * @private
+   */
+  this.scope_ = $scope;
+
+  /**
+   * @type {angular.$compile}
+   * @private
+   */
+  this.compile_ = $compile;
+
+  /**
+   * @type {ngeo.CreateGeoJSONBloodhound}
+   * @private
+   */
+  this.ngeoCreateGeoJSONBloodhound_ = ngeoCreateGeoJSONBloodhound;
+
+  var map = this.scope_['getMapFn']();
+  goog.asserts.assertInstanceof(map, ol.Map);
+
+  /**
+   * @type {!ol.Map}
+   * @private
+   */
+  this.map_ = map;
+
+  /**
+   * @type {ol.layer.Vector}
+   * @private
+   */
+  this.vectorLayer_ = this.createVectorLayer_();
+
+  var datasources = this.scope_['getDatasourcesFn']();
+  goog.asserts.assertArray(datasources);
+
+  /**
+   * @type {Array.<gmfx.SearchDirectiveDatasource>}
+   * @private
+   */
+  this.datasources_ = datasources;
+
+  /**
+   * @type {TypeaheadOptions}
+   * @export
+   */
+  this.options = /** @type {TypeaheadOptions} */ ({
+    highlight: true,
+    hint: undefined,
+    minLength: undefined
+  });
+
+  /**
+   * @type {Array.<TypeaheadDataset>}
+   * @export
+   */
+  this.datasets = [];
+
+  for (var i = 0; i < this.datasources_.length; i++) {
+    var datasource = this.datasources_[i];
+    var groupByProperties = datasource.groupByProperties || [];
+    var filter;
+
+    do {
+      var title = datasource.title;
+
+      if (groupByProperties[0]) {
```

Comment:
I'd use `groupByProperties.lengh > 0` here, to use the same test as in the `while` condition.